src/partition/ReceivedPartition.cpp

```diff
@@ -175,95 +344,125 @@ void ReceivedPartition::compute()
+    std::vector<int>                connectedRanksList; // local ranks with any connection
+
+    // connected ranks for master
+    for (auto &remoteBB : remoteBBMap) {
+      if (overlapping(_bb, remoteBB.second)) {
+        _mesh->getConnectedRanks().push_back(remoteBB.first); //connected remote ranks for this rank
       }
-      PRECICE_DEBUG("Send partition feedback to master");
-      utils::MasterSlave::_communication->send(vertexIDs, 0);
     }
-    int globalNumberOfVertices = -1;
-    utils::MasterSlave::_communication->broadcast(globalNumberOfVertices, 0);
-    PRECICE_ASSERT(globalNumberOfVertices >= 0);
-    _mesh->setGlobalNumberOfVertices(globalNumberOfVertices);
-  } else { // Master
-    int              numberOfVertices = _mesh->vertices().size();
-    std::vector<int> vertexIDs(numberOfVertices, -1);
-    for (int i = 0; i < numberOfVertices; i++) {
-      vertexIDs[i] = _mesh->vertices()[i].getGlobalIndex();
+    if (not _mesh->getConnectedRanks().empty()) {
+      connectionMap[0] = _mesh->getConnectedRanks();
+      connectedRanksList.push_back(0);
     }
-    _mesh->getVertexDistribution()[0] = std::move(vertexIDs);
 
-    for (int rankSlave = 1; rankSlave < utils::MasterSlave::getSize(); rankSlave++) {
-      int numberOfSlaveVertices = -1;
-      utils::MasterSlave::_communication->receive(numberOfSlaveVertices, rankSlave);
-      PRECICE_ASSERT(numberOfSlaveVertices >= 0);
-      std::vector<int> slaveVertexIDs(numberOfSlaveVertices, -1);
-      if (numberOfSlaveVertices != 0) {
-        PRECICE_DEBUG("Receive partition feedback from slave rank " << rankSlave);
-        utils::MasterSlave::_communication->receive(slaveVertexIDs, rankSlave);
+    // receive connected ranks from slaves and add them to the connection map
+    for (int rank = 1; rank < utils::MasterSlave::getSize(); rank++) {
+      std::vector<int> slaveConnectedRanks;
+      int              connectedRanksSize = -1;
+      utils::MasterSlave::_communication->receive(connectedRanksSize, rank);
+      if (connectedRanksSize != 0) {
+        connectedRanksList.push_back(rank);
+        utils::MasterSlave::_communication->receive(slaveConnectedRanks, rank);
+        connectionMap[rank] = slaveConnectedRanks;
       }
-      _mesh->getVertexDistribution()[rankSlave] = std::move(slaveVertexIDs);
     }
-    utils::MasterSlave::_communication->broadcast(_mesh->getGlobalNumberOfVertices());
+
+    // send connectionMap to other master
+    _m2ns[0]->getMasterCommunication()->send(connectedRanksList, 0);
+    if (not connectionMap.empty()) {
+      com::CommunicateBoundingBox(_m2ns[0]->getMasterCommunication()).sendConnectionMap(connectionMap, 0);
+    } else {
+      PRECICE_ERROR("This participant seems to have no mesh partitions for mesh " << _mesh->getName() << " at the coupling interface.");
+    }
+  } else {
+    PRECICE_ASSERT(utils::MasterSlave::isSlave());
+
+    for (auto &remoteBB : remoteBBMap) {
```

Comment:
As you are not modifying its content:
```suggestion
    for (const auto &remoteBB : remoteBBMap) {
```